File: Difficulty: Medium/Minimum Cost to cut a board into squares/minimum-cost-to-cut-a-board-into-squares.cpp
// User function Template for C++

class Solution {
  public:
    int minimumCostOfBreaking(vector<int> x, vector<int> y, int m, int n) {
        // Write your code here
        
        sort(x.begin(),x.end(),greater<int>());
        sort(y.begin(),y.end(),greater<int>());
        int i = 0, j = 0;
        int h = 1, v = 1;
        int ans = 0;
        
        while(i<m-1 && j<n-1) {
            if(x[i]>y[j]) {
                ans = ans + (x[i]*v);
                i++;
                h  = h+1;
            }
            else {
                ans = ans + (y[j]*h);
                j++;
                v = v+1;
            }
        }
        while(i<m-1) {
            ans = ans+(v*x[i]);
            i++;
        }
        while(j<n-1) {
            ans = ans+(h*y[j]);
            j++;
        }
        return ans;
        
    }
};

File: Difficulty: Medium/Minimum Cost to cut a board into squares/minimum-cost-to-cut-a-board-into-squares.java
class Solution {
    public static int minCost(int n, int m, int[] x, int[] y) {
        // code here
        int xs = x.length;
        int ys = y.length;
        int h = 1;
        int v = 1;
        Arrays.sort(x);
        Arrays.sort(y);
        int i = xs-1;
        int j = ys-1;
        int totalsum = 0;
        while(i>=0 && j>=0) {
            int sum = 0;
            if(x[i]>y[j]) {
                sum = (x[i] * h);
                v+=1;
                i--;
            }
            else if(y[j] > x[i]) {
                sum+= (y[j]*v);
                h+=1;
                j--;
            }
            else {
                if(h>v) {
                    sum+= (y[j]*v);
                    h+=1;
                     j--;
                }
                else {
                    sum  += (x[i] * h);
                    v+=1;
                    i--;
                }
            }
            
            //System.out.println(sum + " = " + h + " " + v);
            
            totalsum+=sum;
            
        }
        
        while(i>=0) {
            totalsum+=(x[i]*h);
            v+=1;
            i--;
        }
        
        while(j>=0) {
            totalsum+=(y[j]*v);
            h+=1;
            j--;
        }
        
        return totalsum;
    }
}
